Bowling.java

```diff
@@ -116,28 +120,50 @@ private Frame getCurrentFrame() {
 		return currentFrame;
 	}
 	
-	private Optional<Frame> previousFrame() {
-		if (allFrames.size() > 1) {
-			int nextToLastIndex = allFrames.size() - 2;
+	private Optional<Frame> getFrameBefore(Frame frame) {
+		int currentIndex = allFrames.indexOf(frame);
+		int lastIndex = allFrames.size() - 1;
+
+		if (allFrames.size() > 1 && currentIndex == lastIndex) {
+			int nextToLastIndex = allFrames.indexOf(frame) - 1;
 			return Optional.of(allFrames.get(nextToLastIndex));
+		} else if (allFrames.size() > 2) {
+			int thirdToLastIndex = allFrames.size() - 3;
+			return Optional.of(allFrames.get(thirdToLastIndex));
 		}
```

Comment:
The if checks here don't speak well to me.  Let me see if I can put them in english.

> If there are more than one frame and the currentIndex is the same as the last index,
> then find the index before the given frame and return that frame.

> Else if there are more than two frames
> then return the third to last frame

While the first one does get the frame before the current frame, it does so after checking things that seem unrelated.  Some questions: What's the index of the given frame?  Could that help clarify what is being checked?

The second if clause doesn't even care about what the current frame is and just returns the third to last.